File: Web/ClientApp/src/app/GeographicRegionHierarchy.ts
import { Hierarchy, HierarchyMember } from "./Hierarchy";
import { Guid } from "./CommonDomainObjects";
import { GeographicRegion } from "./Locations";

export interface GeographicRegionHierarchy extends Hierarchy<
    Guid,
    GeographicRegionHierarchy,
    GeographicRegionHierarchyMember,
    GeographicRegion>
{
}

export interface GeographicRegionHierarchyMember extends HierarchyMember<
    Guid,
    GeographicRegionHierarchy,
    GeographicRegionHierarchyMember,
    GeographicRegion>
{
}

File: Geophysical/GeographicRegionHierarchy.cs
﻿using CommonDomainObjects;
using System;
using System.Collections.Generic;

namespace Geophysical
{
    public class GeographicRegionHierarchy: Hierarchy<Guid, GeographicRegionHierarchy, GeographicRegionHierarchyMember, GeographicRegion>
    {
        protected GeographicRegionHierarchy() : base()
        {
        }

        public GeographicRegionHierarchy(
            IDictionary<GeographicRegion, IList<GeographicRegion>> hierachy
            ) : base(
                Guid.NewGuid(),
                hierachy)
        {
        }

        protected override GeographicRegionHierarchyMember NewHierarchyMember(
            GeographicRegion                geographicalRegion,
            GeographicRegionHierarchyMember parentHierarchyMember
            )
        {
            return new GeographicRegionHierarchyMember(
                this,
                geographicalRegion,
                parentHierarchyMember);
        }
    }

    public class GeographicRegionHierarchyMember: HierarchyMember<Guid, GeographicRegionHierarchy, GeographicRegionHierarchyMember, GeographicRegion>
    {
        protected GeographicRegionHierarchyMember() : base()
        {
        }

        internal GeographicRegionHierarchyMember(
            GeographicRegionHierarchy       hierarchy,
            GeographicRegion                geographicRegion,
            GeographicRegionHierarchyMember parent
            ) : base(
                Guid.NewGuid(),
                hierarchy,
                geographicRegion,
                parent)
        {
        }
    }
}
